arangod/Aql/AqlValue.h

```diff
@@ -123,6 +123,47 @@ struct AqlValue final {
   friend struct std::hash<arangodb::aql::AqlValue>;
   friend struct std::equal_to<arangodb::aql::AqlValue>;
 
+ public:
+  /// @brief auxiliary struct to quickly compute hashes of the underyling
+  /// AqlValue. It just hashes the stored pointer, and nothing else (not
+  /// even the type - the idea is that two AqlValues with different types
+  /// will not manage memory via the same pointer). 
+  /// note: only works reliably if the underlying AqlValue has a type other
+  /// than VPACK_INLINE!!! This is the case for AqlValues inserted into the
+  /// _valueCount map of an AqlItemBlock, for example. Here, only AqlValues
+  /// with dynamic memory allocations are inserted, thus the assumption 
+  /// holds true
+  struct SimpleHash {
+    SimpleHash() : _hasher() {}
+
+    inline size_t operator()(AqlValue const& value) const noexcept {
+      TRI_ASSERT(value.type() != VPACK_INLINE);
+      /// hash only the pointer. should be very quick
+      return _hasher(value._data.pointer);
```

Comment:
Afaik this is formally UB if e.g. `docvec` was set instead of `pointer`.

If we want to be painstaking, we'd have to check the type and read the "correct" member. In that case it'd probably also be fine to use `&_data.internal[0]` for inline values as well, would it not?

Reply:
All internal types but the VPACK_INLINE type use a pointer value, which is located in the first 8 bytes of `_data`.
It may be UB to treat all these types alike, so I can change it to hash the first 8 bytes inside `_data` if you think that is better.
That would work for VPACK_INLINE types too, although they will not occur inside the map and the hashing may only touch a part of the VPACK_INLINE value (it may be 1 to 15 bytes long).

Reply:
I have no real concerns that the data will be wrongly interpreted. But I wouldn't eliminate the possibility, at least without further research, that statements like `_data.slice = foo;` and `fun(_data.pointer);` might be reordered, because the read to `.pointer` after the write to `.slice` is UB anyway. It's possibly not happening, as `union`s were used for type punning, so maybe no compiler dares to use that as an optimization opportunity. But I don't know and I'd like to be sure. If you're confident that compilers won't do it, its good enough for me.